README.md

```diff
@@ -4,4 +4,213 @@
   SPDX-License-Identifier: Apache-2.0
 -->
 
-# astarte-message-hub
+# Astarte Message Hub
+
+A central service that runs on (Linux) devices for collecting and delivering messages from N apps using 1 MQTT connection to Astarte.
+
+## Architecture
+The Astarte Message Hub is the main component that shares the Astarte connection to the nodes attached to it.
+The communication between the Hub and nodes is based on `gRPC` which is great for scenarios like real-time communication, low-power, low-bandwidth systems, and multi-language environments.
+`gRPC` makes the most out of `HTTP/2`, with multiplexed streaming and binary protocol framing. In addition, it offers performance advantages through the `Protobuf` message structure and features built-in code generation capability, which enables a multilingual environment.
+
+```mermaid
+flowchart LR
+  subgraph arch[  ]
+    direction TB
+    subgraph cloud[ ]
+        style cloud fill:#00b0f1,stroke:#ffffff
+        direction LR
+        subgraph astarte[Astarte]
+        style astarte fill:#127ab2,stroke:#ffffff,color:#fff
+        end
+    end
+    subgraph device[   ]
+    style device fill:#00b0f1, stroke:#ffffff
+    direction TB
+        astarte-mh[Astarte Message Hub]
+            style astarte-mh fill:#127ab2,stroke:#ffffff,color:#fff
+        astarte-mh <--> node1[Node 1]
+        astarte-mh <--> node2[Node 2]
+        astarte-mh <--> noden[Node...N]
+        style node1 stroke:#ffffff
+        style node2 stroke:#ffffff
+        style noden stroke:#ffffff
+    end
+  end
+  astarte <==> astarte-mh
+  style arch fill:#25516e,stroke:#c4c4c4,stroke-width:1 px,color:#fff,stroke-dasharray: 2 1
+```
+
+## GRPC MessageHub Service
+### Node
+A node is an entity connected to Astarte Message Hub, it can receive/send message from/to Astarte via the Message Hub.
+A node has a uniquely identified by its Node ID and has an Introspection that is a list of Interfaces used by it for exchanging data with Astarte. 
+
+``` protobuf
+message Node {
+  string id = 1;
+  repeated Interface introspection = 2;
+}
+```
+### Interface 
+[Interfaces](https://docs.astarte-platform.org/snapshot/030-interface.html) are a core concept of Astarte which defines how data is exchanged between Astarte and its peers.
+
+``` protobuf
+message Interface {
+  string name = 1;
+  int32 major = 2;
+  int32 minor = 3;
+}
+```
+
+### Attach Method
+When a new node is connected to Message Hub, it will have to call `Attach` method to exchange data with Astarte.
+If the node was successfully attached, the method returns a gRPC stream into which the events received from Astarte(based on the declared Introspection) will be redirected.
+
+``` protobuf
+service MessageHub {
+  rpc Attach(Node) returns (stream AstarteMessage) {}
```

Comment:
This structure doesn't allow to signal errors to the caller. What if, e.g., two different nodes declare two different major versions of the same interface? I think we should separate this between an initial "handshake" phase, where the node sends its information and the server replies with "Ok I identified you" or "There's this error" and then a separate call to attach to the AstarteMessage stream

Reply:
Because in this case, we use directly the gRPC way for error handlig. All the `returns` in a gRPC method provide a Response with a couple of (object, status).
If an error occurs, gRPC returns one of its error status codes instead, with an optional string error message that provides further details about what happened. Error information is available to gRPC clients in all supported languages. [ErrorHandling](https://www.grpc.io/docs/guides/error/)
for example in this case :
``` rust 
RpcStatus::new( RpcStatusCode::InvalidArgument,
Some( "major is different" )
)
```